android/canonical/app/src/main/java/com/google/samples/quickstart/canonical/MainActivity.kt

```diff
@@ -58,5 +128,31 @@ class MainActivity : AppCompatActivity() {
   }
 
 
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    googleSignInInit()
+    setupNavigationBar()
```

Comment:
You might want to check if user has already signed in here with FirebaseAuth.currentUser. And separate the sign in logic from the main activity. If user has signed in before, just continue to setup the navigation. If not, redirect to the login activity.

https://firebase.google.com/docs/auth/web/manage-users#get_the_currently_signed-in_user

Reply:
The check process is implemented in onStart() method. The setupNavigationBar() will be executed anyway. If user hasn't signed in before, layout sign_in_page.xml will be loaded.

As for separating the sign in logic, I adopted this at first and I integrated finally to follow the single-activity-multi-fragments style. Seems google recommend it.
Some links
[Single activity architecture](https://proandroiddev.com/part-3-single-activity-architecture-514791724172#:~:text=Single%20activity%20architecture%20is%20a,which%20now%20require%20less%20boilerplate.)
[Youtube](https://www.youtube.com/watch?v=2k8x8V77CrU)

Of course, using separate activities can achieve the same result and I would like to modify it if you think that is better

Reply:
Sorry for not being clear. I think login stuff should be in its own fragment, instead of putting all the logic inside of main activity.